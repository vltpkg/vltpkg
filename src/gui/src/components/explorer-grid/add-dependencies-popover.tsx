import {
  type ChangeEvent,
  useState,
  type SyntheticEvent,
  type KeyboardEvent,
  type MouseEvent,
} from 'react'
import { BatteryLow, PackageCheck, PackagePlus } from 'lucide-react'
import { Button } from '@/components/ui/button.jsx'
import { CardHeader, CardTitle } from '@/components/ui/card.jsx'
import { Input } from '@/components/ui/input.jsx'
import { Label } from '@/components/ui/form-label.jsx'
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from '@/components/ui/select.jsx'
import { LoadingSpinner } from '@/components/ui/loading-spinner.jsx'

export type InstallOptions = {
  name: string
  version: string
  type: string
}

export type AddDependenciesPopoverProps = {
  error: string
  inProgress: boolean
  onInstall: (o: InstallOptions) => void
  onClose: () => void
}

export const AddDependenciesPopover = ({
  error,
  inProgress,
  onInstall,
  onClose,
}: AddDependenciesPopoverProps) => {
  const [packageName, setPackageName] = useState<string>('')
  const [packageVersion, setPackageVersion] =
    useState<string>('latest')
  const [packageType, setPackageType] = useState<string>('prod')
  const formSubmit = (e: SyntheticEvent) => {
    e.stopPropagation()
    e.preventDefault()
    onInstall({
      name: packageName,
      version: packageVersion,
      type: packageType,
    })
  }
  // we need to add this extra event handler in order to avoid
  // radix-ui/popover from closing the popup on pressing enter
  const keyDown = (e: KeyboardEvent) => {
    if (e.key === 'Enter') {
      formSubmit(e)
    }
  }

  const closePopover = (e: MouseEvent) => {
    e.preventDefault()
    e.stopPropagation()
    onClose()
  }

  if (inProgress) {
    return (
      <>
        <CardHeader className="relative flex w-full flex-col rounded-t-lg p-0">
          <div className="flex items-center justify-between px-6 py-4">
            <CardTitle className="flex items-center text-lg font-medium">
              <PackagePlus size={18} className="mr-2" />
              Add new dependency
            </CardTitle>
          </div>
        </CardHeader>
        <div className="flex h-48 items-center justify-center">
          <LoadingSpinner />
        </div>
      </>
    )
  }

  if (error) {
    return (
      <>
        <CardHeader className="relative flex w-full flex-col rounded-t-lg p-0">
          <div className="flex items-center justify-between px-6 py-4">
            <CardTitle className="flex items-center text-lg font-medium">
              <BatteryLow size={18} className="mr-2" />
              Error
            </CardTitle>
          </div>
        </CardHeader>
        <div className="flex flex-row flex-wrap items-center justify-between gap-2 border-t-[1px] border-muted-foreground/20 p-6">
          <p className="text-sm">{error}</p>
        </div>
        <div className="flex w-full justify-end">
          <Button
            className="m-2"
            role="cancel"
            variant="secondary"
            tabIndex={4}
            onClick={closePopover}>
            Close
          </Button>
        </div>
      </>
    )
  }

  return (
    <>
      <CardHeader className="relative flex w-full flex-col rounded-t-lg p-0">
        <div className="flex items-center justify-between px-6 py-4">
          <CardTitle className="flex items-center text-lg font-medium">
            <PackagePlus size={18} className="mr-2" />
            Add new dependency
          </CardTitle>
        </div>
      </CardHeader>
      <form
        className="flex flex-row flex-wrap items-center justify-between gap-2 border-t-[1px] border-muted-foreground/20 p-6"
        onSubmit={formSubmit}
        onKeyDown={keyDown}>
        <Label htmlFor="package-name" className="ml-1">
          Package Name
        </Label>
        <Input
          id="package-name"
          tabIndex={1}
          type="text"
          role="input"
          placeholder="Package name"
          value={packageName}
          onChange={(e: ChangeEvent) => {
            const value = (e.currentTarget as HTMLInputElement).value
            setPackageName(value)
          }}
        />
        <Label htmlFor="package-version" className="ml-1 mt-2">
          Version | Spec
        </Label>
        <Input
          id="package-version"
          tabIndex={2}
          type="text"
          role="input"
          value={packageVersion}
          onChange={(e: ChangeEvent) => {
            const value = (e.currentTarget as HTMLInputElement).value
            setPackageVersion(value)
          }}
        />
        <Label htmlFor="package-type" className="ml-1 mt-2">
          Type
        </Label>
        <Select
          defaultValue={packageType}
          onValueChange={setPackageType}>
          <SelectTrigger>
            <SelectValue tabIndex={3} />
          </SelectTrigger>
          <SelectContent className="z-[10001]">
            <SelectItem value="prod">dependencies</SelectItem>
            <SelectItem value="dev">devDependencies</SelectItem>
            <SelectItem value="optional">
              optionalDependencies
            </SelectItem>
            <SelectItem value="peer">peerDependencies</SelectItem>
          </SelectContent>
        </Select>
        <div className="flex w-full justify-end">
          <Button
            className="mr-2 mt-2"
            role="cancel"
            variant="secondary"
            tabIndex={4}
            onClick={closePopover}>
            Cancel
          </Button>
          <Button className="mt-2" role="submit" tabIndex={5}>
            <PackageCheck size={16} />
            Install package
          </Button>
        </div>
      </form>
    </>
  )
}
