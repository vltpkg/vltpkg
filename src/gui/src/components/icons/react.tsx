export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
