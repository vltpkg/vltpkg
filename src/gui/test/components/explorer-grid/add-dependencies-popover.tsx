import { vi, test, expect, afterEach } from 'vitest'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { useGraphStore as useStore } from '@/state/index.js'
import { AddDependenciesPopover } from '@/components/explorer-grid/add-dependencies-popover.jsx'

vi.mock('lucide-react', () => ({
  BatteryLow: 'gui-battery-low-icon',
  PackageCheck: 'gui-package-check-icon',
  PackagePlus: 'gui-package-plus-icon',
}))

vi.mock('@/components/ui/button.jsx', () => ({
  Button: 'gui-button',
}))

vi.mock('@/components/ui/card.jsx', () => ({
  CardHeader: 'gui-card-header',
  CardTitle: 'gui-card-title',
}))

vi.mock('@/components/ui/input.jsx', () => ({
  Input: 'gui-input',
}))

vi.mock('@/components/ui/form-label.jsx', () => ({
  Label: 'gui-label',
}))

vi.mock('@/components/ui/select.jsx', () => ({
  Select: 'gui-select',
  SelectContent: 'gui-select-content',
  SelectItem: 'gui-select-item',
  SelectTrigger: 'gui-select-trigger',
  SelectValue: 'gui-select-value',
}))

vi.mock('@/components/ui/popover.jsx', () => ({
  PopoverClose: 'gui-popover-close',
}))

vi.mock('@/components/ui/loading-spinner.jsx', () => ({
  LoadingSpinner: 'gui-loading-spinner',
}))

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

afterEach(() => {
  const CleanUp = () => (useStore(state => state.reset)(), '')
  render(<CleanUp />)
  cleanup()
})

test('add-dependencies-popover render default', async () => {
  render(
    <AddDependenciesPopover
      error=""
      inProgress={false}
      onInstall={() => {}}
      onClose={() => {}}
    />,
  )
  expect(window.document.body.innerHTML).toMatchSnapshot()
})

test('add-dependencies-popover error', async () => {
  render(
    <AddDependenciesPopover
      error="Failed to install package"
      inProgress={false}
      onInstall={() => {}}
      onClose={() => {}}
    />,
  )
  expect(window.document.body.innerHTML).toMatchSnapshot()
})

test('add-dependencies-popover in progress', async () => {
  render(
    <AddDependenciesPopover
      error=""
      inProgress={true}
      onInstall={() => {}}
      onClose={() => {}}
    />,
  )
  expect(window.document.body.innerHTML).toMatchSnapshot()
})
