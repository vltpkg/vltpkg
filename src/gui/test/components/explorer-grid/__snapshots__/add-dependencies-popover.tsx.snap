// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`add-dependencies-popover error 1`] = `

<div>
  <gui-card-header classname="relative flex w-full flex-col rounded-t-lg p-0">
    <div class="flex items-center justify-between px-6 py-4">
      <gui-card-title classname="flex items-center text-lg font-medium">
        <gui-battery-low-icon
          size="18"
          classname="mr-2"
        >
        </gui-battery-low-icon>
        Error
      </gui-card-title>
    </div>
  </gui-card-header>
  <div class="flex flex-row flex-wrap items-center justify-between gap-2 border-t-[1px] border-muted-foreground/20 p-6">
    <p class="text-sm">
      Failed to install package
    </p>
  </div>
  <div class="flex w-full justify-end">
    <gui-button
      classname="m-2"
      role="cancel"
      variant="secondary"
      tabindex="4"
    >
      Close
    </gui-button>
  </div>
</div>

`;

exports[`add-dependencies-popover in progress 1`] = `

<div>
  <gui-card-header classname="relative flex w-full flex-col rounded-t-lg p-0">
    <div class="flex items-center justify-between px-6 py-4">
      <gui-card-title classname="flex items-center text-lg font-medium">
        <gui-package-plus-icon
          size="18"
          classname="mr-2"
        >
        </gui-package-plus-icon>
        Add new dependency
      </gui-card-title>
    </div>
  </gui-card-header>
  <div class="flex h-48 items-center justify-center">
    <gui-loading-spinner>
    </gui-loading-spinner>
  </div>
</div>

`;

exports[`add-dependencies-popover render default 1`] = `

<div>
  <gui-card-header classname="relative flex w-full flex-col rounded-t-lg p-0">
    <div class="flex items-center justify-between px-6 py-4">
      <gui-card-title classname="flex items-center text-lg font-medium">
        <gui-package-plus-icon
          size="18"
          classname="mr-2"
        >
        </gui-package-plus-icon>
        Add new dependency
      </gui-card-title>
    </div>
  </gui-card-header>
  <form class="flex flex-row flex-wrap items-center justify-between gap-2 border-t-[1px] border-muted-foreground/20 p-6">
    <gui-label
      htmlfor="package-name"
      classname="ml-1"
    >
      Package Name
    </gui-label>
    <gui-input
      id="package-name"
      tabindex="1"
      type="text"
      role="input"
      placeholder="Package name"
      value
    >
    </gui-input>
    <gui-label
      htmlfor="package-version"
      classname="ml-1 mt-2"
    >
      Version | Spec
    </gui-label>
    <gui-input
      id="package-version"
      tabindex="2"
      type="text"
      role="input"
      value="latest"
    >
    </gui-input>
    <gui-label
      htmlfor="package-type"
      classname="ml-1 mt-2"
    >
      Type
    </gui-label>
    <gui-select>
      <gui-select-trigger>
        <gui-select-value tabindex="3">
        </gui-select-value>
      </gui-select-trigger>
      <gui-select-content classname="z-[10001]">
        <gui-select-item value="prod">
          dependencies
        </gui-select-item>
        <gui-select-item value="dev">
          devDependencies
        </gui-select-item>
        <gui-select-item value="optional">
          optionalDependencies
        </gui-select-item>
        <gui-select-item value="peer">
          peerDependencies
        </gui-select-item>
      </gui-select-content>
    </gui-select>
    <div class="flex w-full justify-end">
      <gui-button
        classname="mr-2 mt-2"
        role="cancel"
        variant="secondary"
        tabindex="4"
      >
        Cancel
      </gui-button>
      <gui-button
        classname="mt-2"
        role="submit"
        tabindex="5"
      >
        <gui-package-check-icon size="16">
        </gui-package-check-icon>
        Install package
      </gui-button>
    </div>
  </form>
</div>

`;
